agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala

```diff
@@ -50,122 +49,124 @@ object AtumMeasure {
-        .agg(sum(col(aggregatedColumnName)))
-        .collect()(0)(0)
-      val resultValue = if (value == null) "" else value.toString
-      ResultOfMeasurement(resultValue, ResultValueType.String)
+      val resultValue = ds.select(columnExpression).collect()
+      ResultOfMeasurement(Option(resultValue(0)(0)).getOrElse("").toString, resultValueType)
     }
 
+    override def controlColumns: Seq[String] = Seq(controlCol)
     override val resultValueType: ResultValueType.ResultValueType = ResultValueType.String
   }
   object SumOfHashesOfColumn {
-    private [agent] val measureName: String = "hashCrc32"
+    private[agent] val measureName: String = "hashCrc32"
     def apply(controlCol: String): SumOfHashesOfColumn = SumOfHashesOfColumn(measureName, controlCol)
   }
 
-  private def aggregateColumn(
-    ds: DataFrame,
-    measureColumn: String,
-    aggExpression: Column
-  ): String = {
-    val dataType = ds.select(measureColumn).schema.fields(0).dataType
-    val aggregatedValue = dataType match {
+  private def castForAggregation(
+    dataType: DataType,
+    column: Column
+  ): Column = {
+    dataType match {
       case _: LongType =>
         // This is protection against long overflow, e.g. Long.MaxValue = 9223372036854775807:
         //   scala> sc.parallelize(List(Long.MaxValue, 1)).toDF.agg(sum("value")).take(1)(0)(0)
         //   res11: Any = -9223372036854775808
         // Converting to BigDecimal fixes the issue
-        // val ds2 = ds.select(col(measurement.controlCol).cast(DecimalType(38, 0)).as("value"))
-        // ds2.agg(sum(abs($"value"))).collect()(0)(0)
-        val ds2 = ds.select(
-          col(measureColumn).cast(DecimalType(38, 0)).as(valueColumnName)
-        )
-        val collected = ds2.agg(aggExpression).collect()(0)(0)
-        if (collected == null) 0 else collected
+        column.cast(DecimalType(38, 0))
       case _: StringType =>
         // Support for string type aggregation
-        val ds2 = ds.select(
-          col(measureColumn).cast(DecimalType(38, 18)).as(valueColumnName)
-        )
-        val collected = ds2.agg(aggExpression).collect()(0)(0)
+        column.cast(DecimalType(38, 18))
+      case _ =>
+        column
+    }
+  }
+
+  private def handleAggregationResult(dataType: DataType, result: Any): String = {
+    val aggregatedValue = dataType match {
+      case _: LongType =>
+        if (result == null) 0 else result
+      case _: StringType =>
         val value =
-          if (collected == null) new java.math.BigDecimal(0)
-          else collected.asInstanceOf[java.math.BigDecimal]
+          if (result == null) new java.math.BigDecimal(0)
+          else result.asInstanceOf[java.math.BigDecimal]
         value.stripTrailingZeros // removes trailing zeros (2001.500000 -> 2001.5, but can introduce scientific notation (600.000 -> 6E+2)
           .toPlainString // converts to normal string (6E+2 -> "600")
       case _ =>
```

Comment:
For the record, this we might want to revisit and rethink. But not part of this PR. 😉